Controller/Cards/DeleteAction.php

```diff
@@ -0,0 +1,120 @@
+<?php
+namespace Omise\Payment\Controller\Cards;
+
+use Magento\Framework\App\Request\Http;
+
+class DeleteAction extends \Magento\Framework\App\Action\Action
+{
+    const WRONG_REQUEST = 1;
+
+    const WRONG_TOKEN = 2;
+
+    const ACTION_EXCEPTION = 3;
+
+    /**
+     * @var array
+     */
+    private $errorsMap = [];
+
+    /**
+     * @var Magento\Customer\Model\Session
+     */
+    private $customerSession;
+
+    /**
+     * @var Omise\Payment\Model\Customer
+     */
+    private $customer;
+
+    /**
+     * @param Context  $context
+     * @param Session  $customerSession
+     * @param Customer $customer
+     */
+    public function __construct(
+        \Magento\Framework\App\Action\Context $context,
+        \Magento\Customer\Model\Session       $customerSession,
+        \Omise\Payment\Model\Customer         $customer
+    ) {
+        parent::__construct($context, $customerSession);
+        $this->customerSession = $customerSession;
+        $this->customer        = $customer;
+
+        $this->errorsMap = [
+            self::WRONG_TOKEN      => __('No token found.'),
+            self::WRONG_REQUEST    => __('Wrong request.'),
+            self::ACTION_EXCEPTION => __('Deletion failure. Please try again.'),
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function execute()
+    {
+        $request = $this->_request;
+        if (!$request instanceof Http) {
+            return $this->createErrorResponse(self::WRONG_REQUEST);
+        }
+
+        $cardId = $this->getCardID($request);
+
+        if ($cardId === null) {
+            return $this->createErrorResponse(self::WRONG_TOKEN);
+        }
+
+        try {
+            $this->customer->deleteCard($cardId);
+        } catch (\Exception $e) {
+            return $this->createErrorResponse(self::ACTION_EXCEPTION);
+        }
+
+        return $this->createSuccessMessage();
+    }
+
+    /**
+     * @param int $errorCode
+     * @return Magento\Framework\App\ResponseInterface
+     */
+    private function createErrorResponse($errorCode)
+    {
+        $this->messageManager->addErrorMessage(
+            $this->errorsMap[$errorCode]
+        );
+
+        return $this->_redirect('omise/cards/listaction');
```

Comment:
Path feels like it should be a constant - same goes for the other action paths

Reply:
could be.. but most paths are not used as constants in Magento 2. check for instance `Magento_Vault\Controller\DeleteAction::createSuccessMessage`

Reply:
Yeah, but you've repeated it twice right in the same place. Maintainability would be better?